pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,242 @@
+		return errors.AssertionFailedf("key %s must be greater than previous key %s", key.Key, s.prevKey)
+	}
+	// At this point, because of the Reset invariant, we can make the following
+	// assumptions about the new key:
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	lastFile, err := s.maybeDoSizeFlush(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	lastFile.EntryCounts.Add(keyAsRowCount)
+	s.flushedSize += keyAsRowCount.DataSize
+	s.prevKey = append(s.prevKey[:0], key.Key...)
+
+	return nil
+}
+
+// Reset resets the SSTSinkKeyWriter to write to a new span. It ensures that an
+// open BackupManifest_File exists for the new span. It will either open a new
+// BackupManifest_File, or if the new span contiguously extends the last span,
+// it will extend it if it remains under the fileSpanByteLimit. The caller is
+// responsible for ensuring that spans that are reset on do not end mid-row.
+// Any time a new span is being written, Reset MUST be called prior to any
+// WriteKey calls.
+func (s *SSTSinkKeyWriter) Reset(ctx context.Context, newSpan roachpb.Span) error {
+	log.VEventf(ctx, 2, "resetting sink to span %s", newSpan)
+	if err := s.maybeFlushNonExtendableSpans(newSpan); err != nil {
+		return err
+	}
+	if s.out == nil {
+		if err := s.open(ctx); err != nil {
+			return err
+		}
+	}
+	// At this point, we can assume the new span is either contiguous with the
+	// last span or comes after the last span (if it exists).
+	var lastFile *backuppb.BackupManifest_File
+	if len(s.flushedFiles) > 0 {
+		lastFile = &s.flushedFiles[len(s.flushedFiles)-1]
+	}
+
+	if lastFile != nil {
+		s.setMidRowForPrevKey(lastFile.Span.EndKey)
+		if isContiguousSpan(lastFile.Span, newSpan) &&
+			(s.midRow || lastFile.EntryCounts.DataSize < fileSpanByteLimit) {
+			log.VEventf(ctx, 2, "extending span %s to %s", lastFile.Span, newSpan)
+			s.stats.spanGrows++
+			lastFile.Span.EndKey = newSpan.EndKey
+			return nil
+		}
+	}
+
+	s.flushedFiles = append(
+		s.flushedFiles,
+		backuppb.BackupManifest_File{
+			Span: newSpan,
+			Path: s.outName,
+		},
+	)
+	s.prevKey = nil
+	return nil
+}
+
+func (s *SSTSinkKeyWriter) Flush(ctx context.Context) error {
+	if len(s.flushedFiles) > 0 {
+		lastFile := s.flushedFiles[len(s.flushedFiles)-1]
+		// If WriteKey was used, it is possible that the last written key was
+		// mid-row, but because no keys were written after, s.midRow was not updated.
```

Comment:
as i said in comment above, i think Flush() should throw an error if s.midKey is true

Reply:
It does throw an error, but that error is thrown in the underlying `FileSSTSink` `Flush()`.

Reply:
cool. i'm fine with this as is.